### PR TITLE
Feat/5 add mutator and accessor methods for account chart

### PR DESCRIPTION
**Pull request by AlexTheWizardL**
# Pull Request: Add Account Management Methods to CashCtrl Integration

## Description

This pull request introduces new methods for managing accounts in the CashCtrl integration module. The following functionalities have been added:

1. **account_chart**: Retrieves the account chart from a remote CashCtrl instance, formatted to the pyledger schema.
2. **add_account**: Adds a new account to the remote CashCtrl instance.
3. **update_account**: Updates an existing account in the remote CashCtrl instance. (Not yet implemented)
4. **delete_account**: Deletes an account from the remote CashCtrl instance.

Additionally, unit tests have been included to validate the functionality of these methods.

### account_chart

```python
def account_chart(self) -> pd.DataFrame:
    """Retrieves the account chart from a remote CashCtrl instance, formatted to the pyledger schema."""
    accounts = self._client.list_accounts()
    result = pd.DataFrame({
        'account': accounts['number'],
        'currency': accounts['currencyCode'],
        'text': accounts['name'],
        'vat_code': accounts['taxName'],
        'group': accounts['path'],
    })

    return StandaloneLedger.standardize_account_chart(result)
```

### add_account

```python
def add_account(self, account: str, currency: str, text: str, vat_code: str | None = None, group: str | None = None):
    def get_single_record(df, column, value, error_message):
        records = df[df[column] == value]
        if len(records) != 1:
            if len(records) < 1:
                raise ValueError(f"{error_message}: '{value}'.")
            raise ValueError(f"{error_message} '{value}' is duplicated.")
        return records['id'].item()

    currency_data = self._client.get("currency/list.json")['data']
    currency_id = get_single_record(pd.DataFrame(currency_data), 'text', currency, "There is no such currency")

    tax_id = None
    if vat_code:
        tax_data = self._client.list_tax_rates()
        tax_id = get_single_record(pd.DataFrame(tax_data), 'text', vat_code, "There is no VAT code")

    category_id = None
    if group:
        category_data = self._client.list_categories('account')
        category_id = get_single_record(pd.DataFrame(category_data), 'path', group, "There is no such group")

    payload = {
        "number": account,
        "currencyId": currency_id,
        "name": text,
        "taxId": tax_id,
        "categoryId": category_id,
    }

    self._client.post("account/create.json", data=payload)
```

### update_account

```python
def update_account(self, account: str, currency: str, text: str, vat_code: str | None = None, group: str | None = None):
    """
    Not implemented yet
    Updates an existing account in the remote CashCtrl instance.
    Parameters:
        account (str): The account number or identifier to be added.
        currency (str): The currency associated with the account.
        text (str): Additional text or description associated with the account.
        vat_code (str, optional): The VAT code to be applied to the account, if any.
        group (str, optional): The category group to which the account belongs, if any.
    """
    raise NotImplementedError
```

### delete_account

```python
def delete_account(self, account: str, allow_missing: bool = False):
    """Deletes an account from the remote CashCtrl instance."""
    accounts = self._client.list_accounts()
    to_delete = accounts.loc[accounts['number'] == account, 'id']

    if len(to_delete) > 0:
        delete_ids = ",".join(to_delete.astype(str))
        self._client.post('account/delete.json', {'ids': delete_ids})
    elif not allow_missing:
        raise ValueError(f"There is no Account '{account}'.")
```

## Unit Tests

The following unit tests have been added to ensure the correctness of the new methods:

- **test_delete_account_non_existent**: Ensure that deleting a non-existent account does not raise an error when `allow_missing` is set to True.
- **test_add_account**: Test adding a valid account.
- **test_update_account**: Test updating an account with valid inputs.
- **test_delete_account**: Test deleting an existent account.
- **test_delete_non_existent_account_raise_error**: Test deleting a non-existent account should raise an error when `allow_missing` is False.
- **test_update_non_existent_account_raise_error**: Test updating a non-existent account should raise an error.
- **test_update_account_with_not_valid_currency_raise_error**: Test updating an account with a non-valid currency should raise an error.

## ChatGPT review:
<details> <summary>Click for ChatGPT prompt</summary>

### Code Quality and Community Standards

1. **Documentation**: The code is well-documented with clear docstrings explaining the purpose and parameters of each method. This is in line with Python community standards for maintaining readability and ease of understanding.

2. **Error Handling**: The code includes comprehensive error handling, particularly in methods like `add_account` and `update_account`, where specific error messages are provided for different failure conditions. This approach is consistent with best practices in the Python community to provide clear feedback to developers and users.

3. **Use of Pandas**: The code leverages the Pandas library for data manipulation, which is a standard practice in the Python community for handling tabular data. The use of Pandas ensures efficient and readable data processing.

4. **Modular Design**: Each function is designed to perform a single task, adhering to the single responsibility principle. This makes the code easier to maintain and test, aligning with the principles of clean code.

5. **Testing**: The inclusion of unit tests demonstrates a commitment to code quality and reliability. The tests cover various scenarios, including edge cases, which is a standard practice in open-source projects to ensure robustness.

### Comparison to Open Source Projects

- **Alignment with Best Practices**: The implementation aligns well with the practices observed in major open-source projects, such as modularity, comprehensive error handling, and thorough documentation.
  
- **Testing Framework**: The use of `pytest` for unit testing is consistent with its widespread adoption in the Python community for its simplicity and effectiveness in writing concise and readable tests.

- **Code Readability**: The code is clean and readable, following PEP 8 guidelines, which is a hallmark of well-maintained open-source projects.

- **Extensibility**: The design allows for future extensions, such as the implementation of the `update_account` method, demonstrating foresight in planning and scalability.

In summary, the code adheres to Python community standards and compares favorably with practices observed in reputable open-source projects. The emphasis on documentation, error handling, modularity, and testing ensures that the codebase is robust, maintainable, and user-friendly.

</details>

@lasuk Please review the changes and provide feedback or approval for merging.

### CODE REVIEW

File: tests/test_account_chart.py [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+
+# Test updating a non existent account should raise an error.
+def test_update_non_existent_account_raise_error():
+    cashctrl_ledger = CashCtrlLedger()
+    with pytest.raises(ValueError):
+        cashctrl_ledger.update_account(account=7777, currency='CHF',
+            text='test create account', vat_code='MwSt. 2.6%', group='/Anlagevermögen'
+        )
+
+# Test adding an account with not valid vat code should raise an error.
```
Comment (lasuk) [inline tests/test_account_chart.py:84]: not valid *vat code* -> not valid *currency*

File: tests/test_account_chart.py [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+            text='test create account', vat_code='MwSt. 2.6%', group='/Anlagevermögen'
+        )
+
+
+# Test updating an account with not valid vat code should raise an error.
+def test_update_account_with_not_valid_currency_raise_error():
+    cashctrl_ledger = CashCtrlLedger()
+    with pytest.raises(ValueError):
+        cashctrl_ledger.update_account(account=7777, currency='',
+            text='test create account', vat_code='MwSt. 2.6%', group='/Anlagevermögen'
```
Comment (lasuk) [inline tests/test_account_chart.py:98]: I believe the arguments should be something like: currency='USD', vat_code='Non-Existing Tax Code'.

File: tests/test_account_chart.py [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+        'account': 7777,
+        'currency': 'CHF',
+        'text': 'test create account',
+        'vat_code': 'MwSt. 2.6%',
+        'group': '/Anlagevermögen'
+    }
+    cashctrl_ledger.add_account(**new_account)
+    updated_accounts= cashctrl_ledger.account_chart().reset_index()
+    created_accounts = updated_accounts[
+        ~updated_accounts.apply(tuple, 1).isin(initial_accounts.apply(tuple, 1))]
```
Comment (lasuk) [inline tests/test_account_chart.py:28]: Now that I am more familiar with performing set operations on DataFrames, I prefer the following logic:
```
outer_join = pd.merge(initial_accounts, updated_accounts, how='outer', indicator=True)
created_accounts = outer_join[outer_join['_merge'] == "right_only"].drop('_merge', axis = 1)
```
Difference:
- The proposed code does not depend on identical columns nor identical column sequence.
- I find the proposed version easier to read.

Please replace similar occurrences in this unit and in the unit test for VAT.

File: tests/test_account_chart.py [hunk 4]
```diff
@@ -0,0 +1,99 @@
+"""
+Unit tests for account charts accessor and mutator methods.
+"""
+
+import pytest
+from cashctrl_ledger import CashCtrlLedger
+
+# Ensure there is no '2222' account on the remote system
```
Comment (lasuk) [inline tests/test_account_chart.py:8]: 2222 -> 7777

File: tests/test_account_chart.py [hunk 5]
```diff
@@ -0,0 +1,99 @@
+"""
+Unit tests for account charts accessor and mutator methods.
```
Comment (lasuk) [inline tests/test_account_chart.py:2]: Use either 'accounts' or 'account chart' (singular) instead of 'account charts' (plural).

File: tests/test_account_chart.py [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+    assert modified_accounts['currency'].item() == new_account['currency']
+    assert modified_accounts['vat_code'].item() == '<values><de>MwSt. 2.6%</de><en>VAT 2.6%</en><fr>TVA 2.6%</fr><it>IVA 2.6%</it></values>'
+    assert modified_accounts['group'].item() == new_account['group']+'/'+new_account['text']
+
+# Test deleting an existent account.
+def test_delete_account():
+    cashctrl_ledger = CashCtrlLedger()
+    cashctrl_ledger.delete_account(account=7777)
+    updated_accounts = cashctrl_ledger.account_chart()
+
```
Comment (lasuk) [inline tests/test_account_chart.py:67]: I would not insert an empty line here.

File: cashctrl_ledger/ledger.py [hunk 7]
```diff
@@ -53,17 +53,110 @@ def _single_account_balance():
...
+        """Retrieves the account chart from a remote CashCtrl instance, formatted to the pyledger schema."""
+        accounts = self._client.list_accounts()
+        result = pd.DataFrame({
+            'account': accounts['number'],
+            'currency': accounts['currencyCode'],
+            'text': accounts['name'],
+            'vat_code': accounts['taxName'],
+            'group': accounts['path'],
+        })
+
```
Comment (lasuk) [inline cashctrl_ledger/ledger.py:66]: I would not insert an empty line here.

File: cashctrl_ledger/ledger.py [hunk 8]
```diff
@@ -53,17 +53,110 @@ def _single_account_balance():
         """
         raise NotImplementedError
 
-    def account_chart():
+    def account_chart(self) -> pd.DataFrame:
+        """Retrieves the account chart from a remote CashCtrl instance, formatted to the pyledger schema."""
```
Comment (lasuk) [inline cashctrl_ledger/ledger.py:57]: For coherence, I would write
```
"""
Retrieves the account chart from a remote CashCtrl instance, formatted to the pyledger schema.
"""
```
Instead of:
```
"""Retrieves the account chart from a remote CashCtrl instance, formatted to the pyledger schema."""
```

File: cashctrl_ledger/ledger.py [hunk 9]
```diff
@@ -53,17 +53,110 @@ def _single_account_balance():
...
+
+        tax_id = None
+        if vat_code:
+            tax_data = self._client.list_tax_rates()
+            tax_id = get_single_record(pd.DataFrame(tax_data), 'text', vat_code, "There is no VAT code")
+
+        category_id = None
+        if group:
+            category_data = self._client.list_categories('account')
+            category_id = get_single_record(pd.DataFrame(category_data), 'path', group, "There is no such group")
```
Comment (lasuk) [inline cashctrl_ledger/ledger.py:148]: Please replace all occurrences of `get_single_record`with the pattern used for `account_map`.

I am not convinced by the use of `get_single_record`. The interface is not simple and clear, because the function does not fulfil a single purpose. For the benefit of code clarity, I usually avoid this type of glue-code and try to find a clear and succinct way to write out the functionality instead. In our case, we can use the same pattern as with account_map.

Let me compare the two patterns as separate code snippets:
I argue that the first snippet is easier to read.
```
        tax_id = None
        if vat_code:
            tax_data = self._client.list_tax_rates()
            tax_map = tax_data.set_index('text')['id'].to_dict()
            if vat_code not in tax_map:
                raise ValueError(f"VAT code '{vat_code}' does not exist.")
```
```
        def get_single_record(df, column, value, error_message):
            records = df[df[column] == value]
            if len(records) != 1:
                if len(records) < 1:
                    raise ValueError(f"{error_message}: '{value}'.")
                raise ValueError(f"{error_message} '{value}' is duplicated.")
            return records['id'].item()
            
        tax_id = None
        if vat_code:
            tax_data = self._client.list_tax_rates()
            tax_id = get_single_record(pd.DataFrame(tax_data), 'text', vat_code, "There is no VAT code")
```

**Review by lasuk — COMMENTED** [review]
Thank you for submitting well-structured, clear and correctly functioning code. Your efforts in maintaining readability and organization are evident and appreciated. 
Based on this solid foundation, I have highlighted a few areas with potential for further improvement.

File: tests/test_account_chart.py [hunk 10]
```diff
@@ -0,0 +1,99 @@
...
+        )
+
+
+# Test updating an account with not valid vat code should raise an error.
+def test_update_account_with_not_valid_currency_raise_error():
+    cashctrl_ledger = CashCtrlLedger()
+    with pytest.raises(ValueError):
+        cashctrl_ledger.update_account(account=7777, currency='',
+            text='test create account', vat_code='MwSt. 2.6%', group='/Anlagevermögen'
+        )
```
Comment (lasuk) [inline tests/test_account_chart.py:99]: Please add test cases where we add or update an account without a `group` or without a `vat_code`.

File: tests/test_account_chart.py [hunk 11]
```diff
@@ -0,0 +1,122 @@
...
+            text='test create account', vat_code='Non-Existing Tax Code', group='/Anlagevermögen'
+        )
+
+# Test updating an account without VAT code should raise an error.
+def test_update_account_without_vat_raise_error():
+    cashctrl_ledger = CashCtrlLedger()
+    with pytest.raises(ValueError):
+        cashctrl_ledger.update_account(account=7777, currency='USD',
+            text='test create account', vat_code='', group='/Anlagevermögen'
+        )
```
Comment (lasuk) [inline tests/test_account_chart.py:106]: Updating an account without VAT should **not** raise an error. 

To specify 'without VAT code', we pass `vat_code=None`.

`vat_code=''` means that we want an account with a VAT code named as an empty string.

File: tests/test_account_chart.py (hunk 3 above)
Comment (lasuk) [reply tests/test_account_chart.py]: Thanks for implementing in `tests/test_account_chart.py`
Please also replace similar occurrences in `tests/test_vat_code.py`

File: tests/test_account_chart.py [hunk 12]
```diff
@@ -0,0 +1,122 @@
...
+
+# Test updating an account without group should raise an error.
```
Comment (lasuk) [inline tests/test_account_chart.py:108]: Did you mean:
```
# Test updating an account with invalid group should raise an error.
```
Instead of:
```
# Test updating an account without group should raise an error.
```
?

File: tests/test_account_chart.py [hunk 13]
```diff
@@ -0,0 +1,122 @@
...
+            text='test create account', vat_code='MwSt. 2.6%', group='/ABC'
+        )
+
+# Test updating an account without group should raise an error.
+def test_update_account_without_group_raise_error():
+    cashctrl_ledger = CashCtrlLedger()
+    with pytest.raises(ValueError):
+        cashctrl_ledger.update_account(account=7777, currency='USD',
+            text='test create account', vat_code='MwSt. 2.6%', group=''
+        )
```
Comment (lasuk) [inline tests/test_account_chart.py:122]: Updating an account without group should **not** raise an error. 

To specify 'without group', we pass `group=None`.

`group=''` means that we want an account with a group named as an empty string (which is by definition invalid)

**Review by lasuk — COMMENTED** [review]
@AlexTheWizardL: Thanks for your modifications. Back to you with one remaining comment from the initial review plus a few new comments.

File: tests/test_account_chart.py (hunk 11 above)
Comment (lasuk) [reply tests/test_account_chart.py]: Updating an account without VAT should ** **not** ** raise an error.

File: tests/test_account_chart.py (hunk 13 above)
Comment (lasuk) [reply tests/test_account_chart.py]: Updating an account without group should ** **not** ** raise an error.

File: tests/test_account_chart.py (hunk 11 above)
Comment (AlexTheWizardL) [reply tests/test_account_chart.py]: Sorry about the misunderstanding

File: cashctrl_ledger/ledger.py [hunk 14]
```diff
@@ -95,17 +95,109 @@ def _single_account_balance():
...
+        if currency not in currency_map:
+            raise ValueError(f"Currency '{currency}' does not exist.")
+        currency_id = currency_map[currency]
+
+        if vat_code is not None:
+            tax_data = self._client.list_tax_rates()
+            tax_map = tax_data.set_index('text')['id'].to_dict()
+            if vat_code not in tax_map:
+                raise ValueError(f"VAT code '{vat_code}' does not exist.")
+            vat_code = tax_map[vat_code]
```
Comment (lasuk) [inline cashctrl_ledger/ledger.py:182]: Please align with 'tax_id' block in `add_account`.

File: tests/test_account_chart.py [hunk 15]
```diff
@@ -0,0 +1,122 @@
...
+        cashctrl_ledger.update_account(account=1140, currency='USD',
+            text='test create account', vat_code='Non-Existing Tax Code', group='/Anlagevermögen'
+        )
+
+# Test updating an account without VAT code should not raise an error.
+def test_update_account_without_vat_not_raise_error():
+    cashctrl_ledger = CashCtrlLedger()
+    cashctrl_ledger.update_account(account=1140, currency='USD',
+        text='test create account', vat_code=None, group='/Anlagevermögen'
+    )
```
Comment (lasuk) [inline tests/test_account_chart.py:105]: This test should check, `account_chart()` correctly returns the newly added account, similar as the test on lines 16-37. 

The title should read
```
# Test adding an account without VAT code
```

This test does check a usual case, not an edge case. It should be located before edge cases in this unit.

**Review by lasuk — COMMENTED** [review]
Thanks for the changes. Back to you with new comments, @AlexTheWizardL.

**Comment by lasuk** [comment]
I've added commit 40b6adab43 (Polish) to complete and clarify the test suite. Tests work as expected. Ready to merge.
